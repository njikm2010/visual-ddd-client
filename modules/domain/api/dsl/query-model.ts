import * as DSL from './interface';
import * as ViewDSL from '@/modules/domain/domain-design/dsl/dsl';
import { Void } from '@/modules/domain/domain-design/dsl/constants';

import {
  BaseContainer,
  IContainer,
  Node,
  Tree,
  transformMeta,
  transformProperty,
  transformSource,
  transformType,
  Rule,
} from './domain-model';

export class Query extends Node<ViewDSL.QueryDSL> {
  rules: Rule[] = [];

  toDSL(): DSL.QueryDSL {
    const { uuid, title, name, description, meta, properties, source, pagination, result } = this.properties;

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      source: transformSource(source),
      rules: this.rules.map(i => i.toDSL()),
      pagination,
      properties: properties.map(i => transformProperty(i, this.getReference)),
      return: {
        type: result ? transformType(result, this.getReference) : Void,
      },
    };
  }
}

export class DTO extends Node<ViewDSL.DTODSL> {
  toDSL(): DSL.DTODSL {
    const { uuid, title, name, description, meta, properties } = this.properties;

    return {
      uuid,
      title,
      name,
      description,
      meta: transformMeta(meta),
      properties: properties.map(i => transformProperty(i, this.getReference)),
    };
  }
}

export class Container extends BaseContainer implements IContainer {
  queries: Query[] = [];
  dtos: DTO[] = [];

  /**
   * 收集所有有效的节点
   */
  nodesIndexByUUID: Map<string, Node> = new Map();

  constructor(tree: Record<string, Tree>) {
    super();

    this.traverse(tree);
  }

  toDSL(): DSL.QueryModelDSL {
    return {
      queries: this.queries.map(i => i.toDSL()),
      dtos: this.dtos.map(i => i.toDSL()),
    };
  }

  getNodeById(id: string): Node<any> | undefined {
    return this.nodesIndexByUUID.get(id);
  }

  handle(node: Tree, tree: Record<string, Tree>): void {
    const type = node.properties.__node_name__;

    switch (type) {
      case 'query': {
        const properties = node.properties as unknown as ViewDSL.QueryDSL;
        const query = new Query(properties.uuid, properties, this);

        this.nodesIndexByUUID.set(query.id, query);
        this.queries.push(query);

        break;
      }
      case 'dto': {
        const properties = node.properties as unknown as ViewDSL.DTODSL;
        const dto = new DTO(properties.uuid, properties, this);

        this.nodesIndexByUUID.set(dto.id, dto);
        this.dtos.push(dto);

        break;
      }
      case 'rule': {
        const properties = node.properties as unknown as ViewDSL.RuleDSL;
        const rule = new Rule(properties.uuid, properties, this);

        this.nodesIndexByUUID.set(rule.id, rule);

        this.addPostTraverse(() => {
          if (properties.aggregator) {
            const aggregator = this.nodesIndexByUUID.get(properties.aggregator.referenceId) as Query | undefined;
            aggregator?.rules.push(rule);
          }
        });

        break;
      }
    }
  }
}

/**
 * DSL 转换
 */
export function transform(tree: Record<string, Tree>): DSL.QueryModelDSL {
  const container = new Container(tree);

  return container.toDSL();
}
