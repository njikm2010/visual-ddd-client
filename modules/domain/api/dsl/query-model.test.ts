import { QueryModelContainer } from './query-model';

const DATA = {
  __ROOT__: {
    __NODE__: true,
    id: '__ROOT__',
    children: {
      'f0934375-2d49-4bc9-bf05-9e81ff44455d': 1,
      '5cabbc44-31d2-4e8f-a684-0a375aa6cfac': 1,
      '43010e5d-a220-4c8c-a316-89580365b7cf': 1,
      'copy-119847fd-8c27-4d68-8460-f4faa058cf68': 1,
      'copy-e4d09acc-a697-4efe-b395-f30dace34efb': 1,
    },
    properties: { __node_name__: '__ROOT__', __node_type__: 'node', __PROPERTY__: true },
  },
  'f0934375-2d49-4bc9-bf05-9e81ff44455d': {
    __NODE__: true,
    id: 'f0934375-2d49-4bc9-bf05-9e81ff44455d',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'query',
      __node_type__: 'node',
      position: { x: 165, y: 63 },
      uuid: 'f0934375-2d49-4bc9-bf05-9e81ff44455d',
      name: 'Untitled',
      title: '未命名',
      description: '',
      meta: [],
      source: {
        http: { enabled: true },
        rpc: { enabled: true },
        event: { enabled: false, value: '' },
        schedule: { enabled: false, value: '' },
      },
      properties: [
        {
          uuid: 'ee44580e-21a3-4382-add0-78e73a40c205',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'base', name: 'String' },
          access: 'public',
        },
      ],
      result: { type: 'reference', referenceId: '5cabbc44-31d2-4e8f-a684-0a375aa6cfac', name: 'Untitled' },
      pagination: true,
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
  '5cabbc44-31d2-4e8f-a684-0a375aa6cfac': {
    __NODE__: true,
    id: '5cabbc44-31d2-4e8f-a684-0a375aa6cfac',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'dto',
      __node_type__: 'node',
      position: { x: 75, y: 390 },
      uuid: '5cabbc44-31d2-4e8f-a684-0a375aa6cfac',
      name: 'Untitled',
      title: '未命名',
      description: '',
      meta: [],
      implements: [],
      abstract: false,
      properties: [
        {
          uuid: '6087b8e2-a8b9-418d-8f6b-8645a7129b2e',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'reference', referenceId: 'copy-e4d09acc-a697-4efe-b395-f30dace34efb', name: 'Hello' },
          access: 'public',
        },
      ],
      methods: [],
      classProperties: [],
      classMethods: [],
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
  '43010e5d-a220-4c8c-a316-89580365b7cf': {
    __NODE__: true,
    id: '43010e5d-a220-4c8c-a316-89580365b7cf',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'rule',
      __node_type__: 'node',
      position: { x: 90, y: 210 },
      uuid: '43010e5d-a220-4c8c-a316-89580365b7cf',
      name: 'Untitled',
      title: '未命名',
      description: '规则描述',
      meta: [],
      aggregator: { referenceId: 'f0934375-2d49-4bc9-bf05-9e81ff44455d', name: '未命名(Untitled)' },
      zIndex: 2,
      __PROPERTY__: true,
    },
  },
  'copy-119847fd-8c27-4d68-8460-f4faa058cf68': {
    __NODE__: true,
    id: 'copy-119847fd-8c27-4d68-8460-f4faa058cf68',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'rule',
      __node_type__: 'node',
      uuid: 'copy-119847fd-8c27-4d68-8460-f4faa058cf68',
      name: 'Untitled2',
      title: '未命名',
      description: '规则描述',
      meta: [],
      aggregator: { referenceId: 'f0934375-2d49-4bc9-bf05-9e81ff44455d', name: '未命名(Untitled)' },
      zIndex: 2,
      position: { x: 304, y: 210 },
      size: { width: 150, height: 94.125 },
      __PROPERTY__: true,
    },
  },
  'copy-e4d09acc-a697-4efe-b395-f30dace34efb': {
    __NODE__: true,
    id: 'copy-e4d09acc-a697-4efe-b395-f30dace34efb',
    parent: '__ROOT__',
    children: {},
    properties: {
      __node_name__: 'dto',
      __node_type__: 'node',
      uuid: 'copy-e4d09acc-a697-4efe-b395-f30dace34efb',
      name: 'Hello',
      title: '你好',
      description: '',
      meta: [],
      implements: [],
      abstract: false,
      properties: [
        {
          uuid: '6087b8e2-a8b9-418d-8f6b-8645a7129b2e',
          name: 'untitled',
          title: '未命名',
          description: '',
          meta: [],
          type: { type: 'base', name: 'String' },
          access: 'public',
        },
      ],
      methods: [],
      classProperties: [],
      classMethods: [],
      zIndex: 2,
      position: { x: 405, y: 420 },
      size: { width: 200, height: 91.125 },
      __PROPERTY__: true,
    },
  },
};

test('query-model', () => {
  const container = new QueryModelContainer(DATA as any);

  expect(container.toDSL()).toMatchSnapshot();
});
