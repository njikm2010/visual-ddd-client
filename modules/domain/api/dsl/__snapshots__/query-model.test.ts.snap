// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query-model 1`] = `
{
  "dtos": [
    {
      "description": "",
      "meta": {},
      "name": "Untitled",
      "properties": [
        {
          "access": "public",
          "description": "",
          "meta": {},
          "name": "untitled",
          "optional": undefined,
          "title": "未命名",
          "type": "[Hello:copy-e4d09acc-a697-4efe-b395-f30dace34efb]",
          "uuid": "6087b8e2-a8b9-418d-8f6b-8645a7129b2e",
        },
      ],
      "title": "未命名",
      "uuid": "5cabbc44-31d2-4e8f-a684-0a375aa6cfac",
    },
    {
      "description": "",
      "meta": {},
      "name": "Hello",
      "properties": [
        {
          "access": "public",
          "description": "",
          "meta": {},
          "name": "untitled",
          "optional": undefined,
          "title": "未命名",
          "type": "String",
          "uuid": "6087b8e2-a8b9-418d-8f6b-8645a7129b2e",
        },
      ],
      "title": "你好",
      "uuid": "copy-e4d09acc-a697-4efe-b395-f30dace34efb",
    },
  ],
  "queries": [
    {
      "description": "",
      "meta": {},
      "name": "Untitled",
      "pagination": true,
      "properties": [
        {
          "access": "public",
          "description": "",
          "meta": {},
          "name": "untitled",
          "optional": undefined,
          "title": "未命名",
          "type": "String",
          "uuid": "ee44580e-21a3-4382-add0-78e73a40c205",
        },
      ],
      "return": {
        "type": "[Untitled:5cabbc44-31d2-4e8f-a684-0a375aa6cfac]",
      },
      "rules": [
        {
          "description": "规则描述",
          "meta": {},
          "name": "Untitled",
          "title": "未命名",
          "uuid": "43010e5d-a220-4c8c-a316-89580365b7cf",
        },
        {
          "description": "规则描述",
          "meta": {},
          "name": "Untitled2",
          "title": "未命名",
          "uuid": "copy-119847fd-8c27-4d68-8460-f4faa058cf68",
        },
      ],
      "source": [
        {
          "type": "http",
        },
        {
          "type": "rpc",
        },
      ],
      "title": "未命名",
      "uuid": "f0934375-2d49-4bc9-bf05-9e81ff44455d",
    },
  ],
}
`;
